### DISCUSSION TITLE
Web e2e testing with Aspire/Playwright on build pipeline

### DISCUSSION THREAD

**Discussion by mip1983** [microsoft/aspire#4587] [Q&A]
I've been playing around with e2e testing on my Aspire/Blazor app, and have been trying out testing against a locally running aspire instance via 'DistributedApplicationTestingBuilder' (using Playwright/nUnit and Specflow).

I've got that working nicely locally, my tests spin up Aspire and Playwright is able to test against the local endpoint.

I've been having a go at seeing if the tests can run this way on my build pipeline (Azure DevOps agent), but realizing that as I'm using a redis cache, the build agent probably can't spin up redis locally as it won't have docker. I imagine the build agent is probably a container anyway, so would this be a container in a container? Is this something that can work or am I barking up the wrong tree?
(I can run these test post deploy as a smoke test instead).

I tried adding the 'Install Docker CLI' step to my pipeline to no avail. With an ubuntu agent I see errors like: 
`error: net::ERR_NETWORK_CHANGED at https://localhost:35925/Dashboard`

With a windows host I see:
```
System.InvalidOperationException : Command C:\hostedtoolcache\windows\dotnet\packs\Aspire.Hosting.Orchestration.win-x64\8.0.1\tools\dcp.exe info returned non-zero exit code -1
TearDown : BoDi.ObjectContainerException : Interface cannot be resolved: Microsoft.Extensions.Hosting.IHost (resolution path: Aspire.Hosting.DistributedApplication)
StackTrace:    at TechTalk.SpecFlow.Infrastructure.TestExecutionEngine.FireEvents(HookType hookType)
```

**Comment by OBlackfinch** [comment]
How were you able to get this far? Sorry I don't have any answer but I'm trying to get there myself, but mind sharing some code on how you got this bit and I wil try to tackle that next.

**Reply by mip1983** [reply]
As in how I run the aspire host to test against? This specFlow specific but I'm sure it can be adapted to whatever you're using:

```cs
[BeforeFeature(Order = 2)]
public static async Task BeforeFeatureLocalAspireHost(IConfigurationRoot configurationRoot, IObjectContainer container)
{
	if (Config.Environment == TestEnvironment.LocalAspire)
	{
		var appHost = await DistributedApplicationTestingBuilder.CreateAsync<Projects.yourproject_Aspire_AppHost>();
		DistributedApplication app = await appHost.BuildAsync();
		await app.StartAsync();

		container.RegisterInstanceAs(app);

		var uri = app.GetEndpoint("your-web-project", "https");
		var url = uri.ToString();

		Config.BaseUrl = url;
	}
}
```

Then I can point Playwright at that Url on my tests.
